e3db-crypto-android/src/main/java/com/tozny/e3db/crypto/FSKSWrapper.java

```diff
@@ -0,0 +1,163 @@
+                            result.load(null, null);
+                            FileOutputStream out = context.openFileOutput(MOBILE_AUTH_DB_KSTORE, Context.MODE_PRIVATE);
+                            try {
+                                result.store(out, getPerf(context).toCharArray());
+                            } finally {
+                                out.close();
+                            }
+                        }
+
+                        fsKS = result;
+                    }
+                    catch (KeyStoreException | IOException | NoSuchAlgorithmException| CertificateException e) {
+                        throw new RuntimeException(e);
+                    }
+                }
+            }
+        }
+
+        return fsKS;
+    }
+
+    private static void saveFSKS(Context context) throws Exception {
+        if (fsKS != null) {
+            synchronized(keyStoreWriteLock) {
+                try {
+                    OutputStream output = context.openFileOutput(MOBILE_AUTH_DB_KSTORE, Context.MODE_PRIVATE);
+                    try {
+                        fsKS.store(output, getPerf(context).toCharArray());
+                        Log.d(TAG, "Saved keystore.");
+                    } finally {
+                        output.close();
+                    }
+                } catch (KeyStoreException | IOException | NoSuchAlgorithmException | CertificateException e) {
+                    Log.e(TAG, e.getMessage());
+                    throw new RuntimeException(e);
+                }
+            }
+        }
+    }
+
+    private static KeyStore.ProtectionParameter getProtectionParameter(KeyProtection protection, String password) {
+        if (protection.protectionType() == KeyProtection.KeyProtectionType.PASSWORD) {
+            if (password == null || password.trim().length() == 0)
+                throw new IllegalArgumentException("password cannot be blank.");
+
+            return new KeyStore.PasswordProtection(password.toCharArray());
+
+        } else if (Build.VERSION.SDK_INT == Build.VERSION_CODES.JELLY_BEAN && protection.protectionType() == KeyProtection.KeyProtectionType.NONE) {
+            // On API 16, the KeyStore will throw when `setEntry` is called with a null ProtectionParam.
+            // Passing a null password wrapped a PasswordProtection instance has the same effect as no password, however.
+            return new KeyStore.PasswordProtection(null);
+
+        } else {
+            return null;
+
+        }
+    }
+
+    private static void createSecretKeyIfNeeded(Context context, String alias, KeyProtection protection, String password) throws Exception {
+
+        KeyStore keyStore = getFSKS(context);
+
+        if (!keyStore.containsAlias(alias)) {
+
+            // TODO: Lilli, is this right?
+            KeyGenerator keyGen = KeyGenerator.getInstance("AES");
+            SecureRandom random = new SecureRandom();
+            keyGen.init(random);
```

Comment:
No real need to pass a `SecureRandom` instance here. However, you do need to specify that the key should be 256-bits (32 bytes) long:

```
keyGen.init(256);
```